File: mini-claw-code-starter-py/src/mini_claw_code_starter_py/tools/read.py
from __future__ import annotations

from typing import Any

from ..types import ToolDefinition


class ReadTool:
    """Chapter 2: build your first tool."""

    def __init__(self) -> None:
        self._definition = None

    @property
    def definition(self) -> ToolDefinition:
        if self._definition is None:
            raise RuntimeError("ReadTool.new() must initialize self._definition")
        return self._definition

    @classmethod
    def new(cls) -> "ReadTool":
        raise NotImplementedError(
            "Create a ToolDefinition named 'read' with a required 'path' parameter"
        )

    async def call(self, args: Any) -> str:
        raise NotImplementedError(
            "Extract 'path', read the file asynchronously, and return the contents"
        )

File: mini-claw-code-starter-ts/src/tools/read.ts
import { readFile } from "node:fs/promises";

import type { JsonValue, Tool } from "../types";
import { ToolDefinition } from "../types";

export class ReadTool implements Tool {
  readonly toolDefinition: ToolDefinition;

  constructor() {
    this.toolDefinition = ToolDefinition.new(
      "read",
      "Read the contents of a file.",
    ).param("path", "string", "The file path to read", true);
  }

  static new(): ReadTool {
    return new ReadTool();
  }

  definition(): ToolDefinition {
    return this.toolDefinition;
  }

  async call(_args: JsonValue): Promise<string> {
    void readFile;
    throw new Error(
      "TODO(ch2): extract args.path, read the file, and return the contents",
    );
  }
}
